spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitStreamMessageHandler.java

```diff
@@ -0,0 +1,259 @@
+			streamMessage = (com.rabbitmq.stream.Message) requestMessage.getPayload();
+		}
+		else {
+			MessageConverter converter = streamOperations.messageConverter();
+			org.springframework.amqp.core.Message amqpMessage = mapMessage(requestMessage, converter,
+					this.headerMapper, this.headersMappedLast);
+			streamMessage = this.streamOperations.streamMessageConverter().fromMessage(amqpMessage);
+		}
+		future = this.streamOperations.send(streamMessage);
+		handleConfirms(requestMessage, future);
+	}
+
+	private void handleConfirms(Message<?> message, ListenableFuture<Boolean> future) {
+		future.addCallback(bool -> this.successCallback.onSuccess(message),
+				ex -> this.failureCallback.failure(message, ex));
+		if (this.waitForConfirm) {
+			try {
+				future.get(this.confirmTimeout, TimeUnit.MILLISECONDS);
+			}
+			catch (InterruptedException ex) {
+				Thread.currentThread().interrupt();
+				throw new MessageHandlingException(message, ex);
+			}
+			catch (ExecutionException | TimeoutException ex) {
+				throw new MessageHandlingException(message, ex);
+			}
+		}
+	}
+
+	/*
+	 * TODO Copied/modified from MapppingUtils until SI 6.0
+	 */
+	private static org.springframework.amqp.core.Message mapMessage(Message<?> message,
+			MessageConverter converter, AmqpHeaderMapper headerMapper, boolean headersMappedLast) {
+
+		MessageProperties amqpMessageProperties = new StreamMessageProperties();
+		org.springframework.amqp.core.Message amqpMessage;
+		if (!headersMappedLast) {
+			mapHeaders(message.getHeaders(), amqpMessageProperties, headerMapper);
+		}
+		if (converter instanceof ContentTypeDelegatingMessageConverter && headersMappedLast) {
+			String contentType = contentTypeAsString(message.getHeaders());
+			if (contentType != null) {
+				amqpMessageProperties.setContentType(contentType);
+			}
+		}
+		amqpMessage = converter.toMessage(message.getPayload(), amqpMessageProperties);
+		if (headersMappedLast) {
+			mapHeaders(message.getHeaders(), amqpMessageProperties, headerMapper);
+		}
+		return amqpMessage;
+	}
+
+	private static void mapHeaders(MessageHeaders messageHeaders, MessageProperties amqpMessageProperties,
+			AmqpHeaderMapper headerMapper) {
+
+		headerMapper.fromHeadersToRequest(messageHeaders, amqpMessageProperties);
+	}
+
+	private static String contentTypeAsString(MessageHeaders headers) {
+		Object contentType = headers.get(AmqpHeaders.CONTENT_TYPE);
+		if (contentType instanceof MimeType) {
+			contentType = contentType.toString();
+		}
+		if (contentType instanceof String) {
+			return (String) contentType;
+		}
+		else if (contentType != null) {
+			throw new IllegalArgumentException(AmqpHeaders.CONTENT_TYPE
+					+ " header must be a MimeType or String, found: " + contentType.getClass().getName());
+		}
+		return null;
+	}
+	/*
+	 * End copied/modified from MappingUtils
+	 */
+
+	@Override
+	public void start() {
+	}
+
+	@Override
+	public void stop() {
+		this.streamOperations.close();
```

Comment:
Will it work after restarting the handler?
I mean since you don't do anything in the `start()` might be this one should go to `destroy()` at all?
Also this doesn't feel kosher to mutate that externally injected `StreamOperations`. Plus it might be stopped outside of this handler and independently.

Do I miss some hidden logic in this approach?

Thanks

Reply:
Yes; all `close()` does is close and null the producer; a new producer is created on the first operation if null.